architecture/2025-02-user-defined-resource-type-feature-spec.md

```diff
@@ -0,0 +1,1361 @@
+
+**User Experience** 
+
+The platform engineer creates a resource type which cannot have a recipe registered for it.
+
+```bash
+# Create a recipe-less resource type just like other resource types
+rad resource-type create --from-file external-service-resource-type.bicep
+Creating resource type MyCompany.App/externalService
+The resource type MyCompany.App/externalService has been created
+```
+
+**`external-service-resource-type.bicep`**:
+
+```yaml
+resource MyCompany.App/externalService 'System.Resources/resourceTypes@2023-10-01-preview' = {
+  name: 'MyCompany.App/externalService'
+  description: '''
+    The external service resource type represents a resource which
+    is not managed by Radius but appears in the application graph.
+  '''
+  // If true (default) allow recipes to be registered to deploy resources
+  // If false, resource can be created in any environment
+  allowRecipes: false 
+  api: {
+    version: 'v1alpha1'
+    schema: {
+      properties: {
+        connection-string: {
+          type: 'string'
+          description: 'The connection string to the external service'
+          connected-resource-environment-variable: EXTERNAL_SERVICE_CONNECTION_STRING
+        }
+        credentials: {
+          type: 'object'
+          properties: {
+            username: {
+              type: 'string'
+              description: 'Username for the external service'
+              connected-resource-environment-variable: EXTERNAL_SERVICE_USERNAME
+            }
+            password: {
+              type: 'string'
+              description: 'Password for the external service user'
+              connected-resource-environment-variable: EXTERNAL_SERVICE_PASSSWORD
+            }
+          }
+      }
+  }
+}
+```
+
+> [!NOTE]
+>
+> This example stores the username and password as cleartext in the Radius database. In a real example, the credentials would be stored in a secret. That was omitted from this example for simplicity. The next user story discusses child resources which could use used to embed a secret in this resource type.
+
+The platform engineer, or environment manager can then create a resource representing the external resource in the environment. For example, the platform engineer may create a resource representing a Twilio account.
+
+**`production-environment.bicep`**
+
+```yaml
+extension radius
+
+resource environment 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'production'
+  ...
+}
+
+var twilio-account-sid = 'ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
+var twilio-username = 'twilio-prod-user'
+var twilio-password = 'o84nouvTiHWiw97sbq6B'
+
+resource twilio 'MyCompany.App/externalService@v1alpha1' = {
+  name: 'twilio'
+  properties: {
+    environment: production
+    connection-string 'https://api.twilio.com/2010-04-01/Accounts/${twilio-account-sid}'
+    credentials:
+      username: ${twilio-username}
+      password: ${twilio-password}
+}
```

Comment:
By default, the resourceProvisioning property gets set to `recipe`. But you can also set it to `manual`: https://github.com/radius-project/radius/blob/main/test/testrecipes/test-bicep-recipes/resource-creation.bicep#L14. This is just an example.